### PR TITLE
add hooks

### PR DESCRIPTION
**Pull request by murad897**

### CODE REVIEW

File: components/List/index.tsx [hunk 1]
```diff
@@ -0,0 +1,28 @@
+import React, { useContext, useEffect } from "react";
+import NavigationContext from "../../context";
```
Comment (d-vorobyov) [inline components/List/index.tsx]: setup `paths` in tsconfig

File: services/index.ts [hunk 2]
```diff
@@ -0,0 +1,16 @@
+import { productsDataProps } from "../types";
+import { productDataProps } from "../types";
+
+export const getProducts = async ({ context, limit, offset }: productsDataProps) => {
+  const response = await fetch(`http://localhost:4000/products?_page=${offset}&_limit=${limit}`, {
```
Comment (d-vorobyov) [inline services/index.ts]: use axios instead

Comment (d-vorobyov) [reply services/index.ts]: it allows to use relative url

File: context/index.tsx [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+  useEffect(() => {
+    axios
+      .get(`http://localhost:4000/products?_page=${offset}&_limit=${limit}`)
+      .then((response) => {
+        setData(response.data);
+      })
+      .catch((err) => console.log(err));
+  }, [offset]);
+
+  const prevHandler = () => {
```
Comment (d-vorobyov) [inline context/index.tsx]: just `prev` and `next` is fine

File: context/index.tsx [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+import { useState, useEffect } from "react";
+import axios from "axios";
+import { ContextInterface } from "../types";
+import { ContextDefaultValues } from "../constants";
+import { Props } from "next/script";
+
+export const NavigationContext = createContext<ContextInterface>(ContextDefaultValues);
+
+export const NavigationProvider = ({ children }: Props) => {
+  const [offset, setOffset] = useState(1);
```
Comment (d-vorobyov) [inline context/index.tsx]: `offset` is not the same as `page`

Comment (d-vorobyov) [reply context/index.tsx]: offset = (page - 1) * limit

Comment (d-vorobyov) [reply context/index.tsx]: page 1 - offset 0
page 2 - offset 5

Comment (d-vorobyov) [reply context/index.tsx]: so rename it

File: context/index.tsx [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+  const [data, setData] = useState([]);
+  const [limit, setLimit] = useState(5);
+  const [paginationActive, setPaginationActive] = useState(false);
+
+  useEffect(() => {
+    axios
```
Comment (d-vorobyov) [inline context/index.tsx]: use axios instance with preconfigured `baseUrl` instead

File: context/index.tsx [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+      .get(`http://localhost:4000/products?_page=${offset}&_limit=${limit}`)
```
Comment (d-vorobyov) [inline context/index.tsx]: pass `_page` and `_limit` as `params` in second arg of `get()`

File: context/index.tsx [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+import { ContextDefaultValues } from "../constants";
+import { Props } from "next/script";
+
+export const NavigationContext = createContext<ContextInterface>(ContextDefaultValues);
+
+export const NavigationProvider = ({ children }: Props) => {
+  const [offset, setOffset] = useState(1);
+  const [data, setData] = useState([]);
+  const [limit, setLimit] = useState(5);
+  const [paginationActive, setPaginationActive] = useState(false);
```
Comment (d-vorobyov) [inline context/index.tsx]: use `is` and `has` prefixes for boolean flags - `isPaginationActive`, `setIsPaginationActive`

File: context/index.tsx [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+import axios from "axios";
+import { ContextInterface } from "../types";
+import { ContextDefaultValues } from "../constants";
+import { Props } from "next/script";
+
+export const NavigationContext = createContext<ContextInterface>(ContextDefaultValues);
+
+export const NavigationProvider = ({ children }: Props) => {
+  const [offset, setOffset] = useState(1);
+  const [data, setData] = useState([]);
```
Comment (d-vorobyov) [inline context/index.tsx]: set data type in `useState<Type[]>([])`

File: context/index.tsx [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+  const [limit, setLimit] = useState(5);
+  const [paginationActive, setPaginationActive] = useState(false);
+
+  useEffect(() => {
+    axios
+      .get(`http://localhost:4000/products?_page=${offset}&_limit=${limit}`)
+      .then((response) => {
```
Comment (d-vorobyov) [inline context/index.tsx]: move request function outside `useEffect` and use `async/await` syntax

File: context/index.tsx [hunk 10]
```diff
@@ -0,0 +1,42 @@
...
+  };
+
+  const nextHandler = () => {
+    if (data.length === 5) {
+      setOffset((prev) => prev + 1);
+      setPaginationActive(true);
+    }
+  };
+
+  return <NavigationContext.Provider value={{ data, paginationActive, prevHandler, nextHandler }}>{children}</NavigationContext.Provider>;
```
Comment (d-vorobyov) [inline context/index.tsx]: too long

File: hooks/UseAuth.ts [hunk 11]
```diff
@@ -0,0 +1,33 @@
+import { useEffect } from "react";
+import { useRouter } from "next/router";
+import { useState } from "react";
```
Comment (d-vorobyov) [inline hooks/UseAuth.ts]: double import from "react"

File: hooks/UseAuth.ts [hunk 12]
```diff
@@ -0,0 +1,33 @@
...
+
+export const UseAuth = () => {
+  const router = useRouter();
+  const [token, setToken] = useState(false);
```
Comment (d-vorobyov) [inline hooks/UseAuth.ts]: what type of this state? string? boolean?

File: hooks/UseLogin.ts [hunk 13]
```diff
@@ -0,0 +1,36 @@
+import axios from "axios";
+import { useRouter } from "next/router";
+import { useState } from "react";
+import { RegisterProps } from "../types";
+
+export const UseLogin = () => {
+  const router = useRouter();
+  const [errorText, setErrorText] = useState("");
```
Comment (d-vorobyov) [inline hooks/UseLogin.ts]: `errorMessage` is more common name

File: hooks/UseLogin.ts [hunk 14]
```diff
@@ -0,0 +1,36 @@
...
+
+  const login = ({ email, password }: RegisterProps) => {
+    try {
+      axios
```
Comment (d-vorobyov) [inline hooks/UseLogin.ts]: instance

File: hooks/UseLogin.ts [hunk 15]
```diff
@@ -0,0 +1,36 @@
...
+        .post(`${process.env.url}/login`, {
+          email: email,
+          password: password,
+        })
+        .then((res) => {
```
Comment (d-vorobyov) [inline hooks/UseLogin.ts]: async/await

File: hooks/UseLogin.ts [hunk 16]
```diff
@@ -0,0 +1,36 @@
...
+          if (res.status === 200 || 201) {
```
Comment (d-vorobyov) [inline hooks/UseLogin.ts]: it doesn't work as you expect) This statement is always `true` since `!!201` is `true`

File: hooks/UseLogin.ts [hunk 17]
```diff
@@ -0,0 +1,36 @@
...
+            router.push("/");
+            localStorage.setItem("token", res.data.accessToken);
+            localStorage.setItem("name", res.data.user.name);
+          }
+        })
+        .catch((err) => {
+          setErrorText(err.message);
+          setTimeout(() => {
```
Comment (d-vorobyov) [inline hooks/UseLogin.ts]: Why?

File: hooks/Useregister.ts [hunk 18]
```diff
@@ -0,0 +1,37 @@
...
+import { useState } from "react";
+import { RegisterProps } from "../types";
+
+export const useRegister = () => {
+  const [errorText, setErrorText] = useState("");
+  const router = useRouter();
+  
+  const register = async ({ name, email, password }: RegisterProps) => {
+    try {
+      axios
```
Comment (d-vorobyov) [inline hooks/Useregister.ts]: instance

File: hooks/Useregister.ts [hunk 19]
```diff
@@ -0,0 +1,37 @@
...
+        .post(`${process.env.url}/register`, {
+          name: name,
+          email: email,
+          password: password,
+        })
+        .then((res) => {
```
Comment (d-vorobyov) [inline hooks/Useregister.ts]: async await

File: hooks/Useregister.ts [hunk 20]
```diff
@@ -0,0 +1,37 @@
...
+          if (res.status === 200 || 201) {
```
Comment (d-vorobyov) [inline hooks/Useregister.ts]: fix

File: hooks/Useregister.ts [hunk 21]
```diff
@@ -0,0 +1,37 @@
...
+          setTimeout(() => {
+            setErrorText("");
+          }, 2000);
+        });
+    } catch (err) {
+      console.log(err);
+    }
+  };
+
+  return { register, errorText };
```
Comment (d-vorobyov) [inline hooks/Useregister.ts:27]: and think, maybe we can use one hook for login and register? code of both seems very-very similar

File: next.config.js [hunk 22]
```diff
@@ -2,6 +2,21 @@
...
+    url: "http://localhost:4000",
+  },
+  images: {
+    domains: ["localhost", "picsum.photos"],
+  },
+  async redirects() {
+    return [
+      {
+        source: "/api/:path*",
+        destination: "http://localhost:4000",
```
Comment (d-vorobyov) [inline next.config.js]: process.env

File: services/index.ts [hunk 23]
```diff
@@ -0,0 +1,16 @@
...
+    headers: { Authentication: context.req.headers.Authentication },
+  });
+  return await response.json();
+};
+
+export const getProduct = async ({ context, id }: productDataProps) => {
+  const response = await fetch(`http://localhost:4000/products/${id}`, {
+    headers: { Authentication: context.req.headers.Authentication },
```
Comment (d-vorobyov) [inline services/index.ts]: it's better to check if `Authentication` provided, and if not - return `{ redirect: "/login" }`

File: services/index.ts [hunk 24]
```diff
@@ -0,0 +1,16 @@
...
+
+export const getProducts = async ({ context, limit, offset }: productsDataProps) => {
+  const response = await fetch(`http://localhost:4000/products?_page=${offset}&_limit=${limit}`, {
+    headers: { Authentication: context.req.headers.Authentication },
+  });
+  return await response.json();
+};
+
+export const getProduct = async ({ context, id }: productDataProps) => {
+  const response = await fetch(`http://localhost:4000/products/${id}`, {
```
Comment (d-vorobyov) [inline services/index.ts]: use axios

File: services/index.ts [hunk 25]
```diff
@@ -0,0 +1,16 @@
...
+    headers: { Authentication: context.req.headers.Authentication },
+  });
+  return await response.json();
```
Comment (d-vorobyov) [inline services/index.ts]: check result before return, app could crash here if response is bad

File: types/index.ts [hunk 26]
```diff
@@ -0,0 +1,43 @@
+export interface RegisterProps {
+  name?: string;
+  email: string;
+  password: string | number;
+}
+export interface listItem {
```
Comment (d-vorobyov) [inline types/index.ts]: **L**istItem

File: types/index.ts [hunk 27]
```diff
@@ -0,0 +1,43 @@
+export interface RegisterProps {
+  name?: string;
+  email: string;
+  password: string | number;
```
Comment (d-vorobyov) [inline types/index.ts]: emmm, no, string only

File: types/index.ts [hunk 28]
```diff
@@ -0,0 +1,43 @@
...
+  cost: number;
+  description: string;
+  familyId: number;
+  id: number;
+  imageUrl: string;
+  locationId: number;
+  name: string;
+  quantity: number;
+}
+export interface listItemComponentProps {
```
Comment (d-vorobyov) [inline types/index.ts]: You can keep props interface in component's file, it's easier to use.

Comment (d-vorobyov) [reply types/index.ts]: Keep here just common data types and interfaces

**Review by d-vorobyov — CHANGES_REQUESTED** [review]

File: components/List/index.tsx (hunk 1 above)
Comment (murad897) [reply components/List/index.tsx]: done

File: hooks/UseLogin.ts (hunk 17 above)
Comment (murad897) [reply hooks/UseLogin.ts]: for skileton animation and error text validation message

**Pull request merged**